Add file browser

drybalka opened the pull request:
This is a minimal implementation of the file browser, which would probably cover a lot of requirements in #200. The whole thing works analogous to the https://github.com/nvim-telescope/telescope-file-browser.nvim as suggested in [this comment](https://github.com/helix-editor/helix/issues/200#issuecomment-1340815804). Even though the resolution of the discussion seems to be "file tree/browser is too hard, it should be implemented as a plugin", I feel like my changes are quite small and natural to be considered for adding to the core nonetheless.

The implementation simply builds on the existing file picker and only modifies 3 files, so the added maintenance burden should be quite small. The code itself is not particularly elegant (in my rather inexperienced opinion), but I did not want to over-complicate things. This is also the reason why some features might be missing.

Feel free to modify this PR or simply make suggestions, I'd be happy to improve it. This is also my first PR here, so sorry if I miss anything.

daedroza commented:
Would it be possible to implement a file browser with this methodology instead? https://github.com/stevearc/oil.nvim

It uses a buffer/pop to navigate and edit files like you're inside a buffer.

gj1118 commented:
@drybalka this is awesome.. thanks for the effort. Can you please post a screenshot as to how it looks ? Basically I am interested to show how nested directories/files are being presented. 
Thanks

drybalka commented:
> Would it be possible to implement a file browser with this methodology instead? https://github.com/stevearc/oil.nvim
> 
> It uses a buffer/pop to navigate and edit files like you're inside a buffer.

I am not a maintainer of helix, but in my opinion this is rather a plugin functionality. First of all, it would be hard-ish to implement and therefore to maintain. Secondly, buffers are primarily used for text editing and one does not usually need to create/delete/rename files so much.

archseer commented:
I kind of like just how simple this change is! It reuses existing UI components and allows exploring the file tree without adding any of the heavier features.

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -265,6 +266,69 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
+    .with_preview(|_editor, path| Some((path.as_path().into(), None)));
+    let injector = picker.injector();
+
+    if let Ok(files) = directory_content {
+        for file in files {
+            if injector.push(file).is_err() {
+                break;
+            }
+        }
+    }
```
Also I believe this function should return a result and pass up the error from `directory_content`. Currently if you fail to list the directory contents a blank picker opens

drybalka replied:
Great suggestions, simplified the code, thank you!

the-mikedavis commented on helix-term/src/commands.rs:
```diff
@@ -337,6 +337,7 @@ impl MappableCommand {
         file_picker, "Open file picker",
         file_picker_in_current_buffer_directory, "Open file picker at current buffer's directory",
         file_picker_in_current_directory, "Open file picker at current working directory",
+        file_browser, "Open file browser at current buffer's directory",
```
```suggestion
        file_browser, "Open file browser at current working directory",
```

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -265,6 +266,69 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
+    picker
+}
+
+fn directory_content(path: &Path) -> Result<Vec<PathBuf>, std::io::Error> {
```
To be able to ascend up a directory we could add `..` to this list whenever the path isn't the root

drybalka replied:
Nice idea and was quite easy to implement too, thanks! Should have added it from the start, really =)

zegervdv commented:
I think it would be useful to have the equivalent version of `file_picker_in_current_buffer_directory`. I often use this, only to realize I need to go one or more levels up. The file_browser would solve that perfectly :).

It could be something like this:
``` rust
fn file_browser_in_current_buffer_directory(cx: &mut Context) {
    let doc_dir = doc!(cx.editor)
        .path()
        .and_then(|path| path.parent().map(|path| path.to_path_buf()));

    let path = match doc_dir {
        Some(path) => path,
        None => {
            cx.editor.set_error("current buffer has no path or parent");
            return;
        }
    };

    if let Ok(picker) = ui::file_browser(path) {
        cx.push_layer(Box::new(overlaid(picker)));
    }
}
```

drybalka commented:
> I think it would be useful to have the equivalent version of `file_picker_in_current_buffer_directory`. I often use this, only to realize I need to go one or more levels up. The file_browser would solve that perfectly :).

I think this can even be the default behavior, thanks for suggesting! Not sure when I would even need to open a file browser in the project root. I deliberately wanted to keep this PR simple and feature-poor, but your suggestion is quite simple and I think it is worth it.

baldwindavid commented:
@drybalka This is perfect with the change to the current buffer directory. It's similar to good ol' netrw for quick navigation and covers 95% of the needs for me. The more advanced stuff (copy, paste, create, delete, etc) is covered by a mix of #11164, shell scripts, and the yazi file explorer. Thanks for your work!

zegervdv commented:
@drybalka found a small issue when testing: if the cursor is on a binary file, the preview will mess up the view and leave random characters everywhere. Maybe binary files can be excluded from preview somehow?
These files are normally excluded from the file_picker, so maybe this is a more general issue with the preview.

drybalka commented:
> @drybalka found a small issue when testing: if the cursor is on a binary file, the preview will mess up the view and leave random characters everywhere. Maybe binary files can be excluded from preview somehow? These files are normally excluded from the file_picker, so maybe this is a more general issue with the preview.

I am using the same file previewer as file_picker, so the problem is probably in the previewer itself. Although as far as I tested the usual .jpg, .pdf, and executables are all correctly previewed as `<Binary file>`. Maybe you're using some exotic file formats?

thomasaarholt commented:
It would be nice to add support for going straight to the root of the project if the current buffer hasn't been saved yet. I went to test the file browser functionality by escaping the file _picker_ when calling `hx .`, but then it didn't work:

https://github.com/user-attachments/assets/22113048-2d31-4382-acd0-0cc5d43ee9d7

thomasaarholt commented:
Actually, I can't get this to work at all? Running on MacOS.

https://github.com/user-attachments/assets/c787a6eb-03bc-44a1-9e71-2a044a49a5c8

[Here is the log](https://gist.github.com/thomasaarholt/fd8de460bb7bf8d028bd7038421ae6d1) from `hx -vvv .`

drybalka commented:
> Actually, I can't get this to work at all? Running on MacOS.

Well, this is actually the behavior I get when using the command palette for the file_picker as well, I guess the palette is somehow buggy in this regard.

L-Trump commented:
> Actually, I can't get this to work at all? Running on MacOS.
> Screen.Recording.2024-09-10.at.08.47.35.mov
> 
> [Here is the log](https://gist.github.com/thomasaarholt/fd8de460bb7bf8d028bd7038421ae6d1) from `hx -vvv .`

Actually in current latest version of helix, no file picker can open from the command pallete (tried in nixos and archlinux). It seems like another bug.

summersz commented:
I have the same reported issue when trying to open the browser from the command palette.  Works great when opened from a keymap though.

I do have a couple of suggestions to consider
 
 1. Can directory names be appended with a '/' to distinguish them (like netrw)
 
 2. Can left and right keys be used to navigate down and up directories, respectively? ( I have gotten used to this in yazi and find it very intuitive)

m0ar commented:
So stoked to see this @drybalka! :heart_hands:

Probably not worth it at this stage as merging the base feature is higher priority, but it'd be cool be be able to classify the entries with icons. Having a trailing slash on dirs makes sense, but hugely useful IMO to when something is linked, and being able to visually filter on filetypes.

Kinda like [`lsd`](https://github.com/lsd-rs/lsd?tab=readme-ov-file#icon-theme-file-content), which has a default set of unicode chars, but being able to opt in to using nerdfont glyphs for extra gloss.

![image](https://github.com/user-attachments/assets/2427ad0b-b92d-42b8-b9ca-eef72654685d)


I could probably take a look at doing this when this is in if you aren't feeling it :)

drybalka commented:
@summersz Adding slash to dirs is a great idea, thank you! Using other keys in a picker is harder. The current design does not allow picker-specific keymaps, and even remapping the existing keys is not allowed yet. I will leave this idea for the future, as pickers refactor should come at some point.

@m0ar I also really like the idea with icons! The file_browser picker shares a lot of code with the file_picker, so if the icons would work there then most probably they will also work here. I don't want to complicate this PR with icons at this point, but you may already try implementing them for the file_picker.

the-mikedavis commented on helix-term/src/ui/picker.rs:
```diff
@@ -276,30 +285,6 @@ pub struct Picker<T: 'static + Send + Sync, D: 'static> {
 }
 
 impl<T: 'static + Send + Sync, D: 'static + Send + Sync> Picker<T, D> {
-    pub fn stream(
```
I'm not sure we will end up using these `stream`+`with_stream` helpers again (I believe #9647 refactored out the only caller) but I think it's premature to remove them now and it's an unrelated change, so let's reset these deletions

drybalka replied:
I reverted these deletions.

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -265,6 +266,76 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
 }
 
+pub fn file_browser(root: PathBuf) -> Result<FilePicker, std::io::Error> {
+    let root = root.canonicalize()?;
```
```suggestion
    let root = helix_stdx::path::canonicalize(root);
```

generally we try to avoid the `canonicalize` from `std` since it resolves symlinks

drybalka replied:
I just copied the code from the file_picker here =) But you are right, we don't want to resolve symlinks in the file browser.

thomasaarholt commented:
To use this, you can add the following keybinding to your config:
```toml
[keys.normal.space.space]
f = "file_browser"
```

Now, you can double-tap `space` and then press `f` to open the file browser.

Note that you cannot open the file browser **from the command palette** due to #4508 (as discussed above).

Akselmo commented:
I love this feature. Could it also change the working-directory too, if for example i navigate to completely different project? Though that is probably for something like `autochdir`

the-mikedavis commented:
That's https://github.com/helix-editor/helix/discussions/10215

Akselmo commented:
> To use this, you can add the following keybinding to your config (`:config-open`, and then after saving `:config-reload`):
> 
> ```toml
> [keys.normal.space.space]
> f = "file_browser"
> ```
> 
> Now, you can double-tap `space` and then press `f` to open the file browser.
> 
> Note that you cannot open the file browser **from the command palette** due to #4508 (as discussed above).

Alternatively if you want to use the file_browser as default, you can replace `file_picker` with `file_browser` in `helix-term/src/keymap/default.rs`

ryanabx reviewed (approved):
Simple and relatively clean impl. Also tested it, and other than https://github.com/helix-editor/helix/issues/4508, which is a separate issue, looks good to me!

It may just be personal preference to me, but I actually preferred having it open in the root of the cwd, as opposed to the directory of the currently open buffer. I'd be fine having any hierarchical file browser over none though, so that's not a blocker for me! Perhaps in the future, having an "open file browser at current working directory" would be a nice-to-have.

It may also be worth considering adding a default keybinding for the file browser, but that might be worth doing in a separate PR to not prolong this one

peteringram0 commented:
I tested your branch and it solves a large problem for me in a very large repo!! I was planning to recommend using the right/left arrows on dir's like @summersz comment but accept your reply on this matter. Im not sure why but there is a flicker when navigating directories, its almost like the window opens and closes quickly which looks jaring but apart form that I think you got a really nice branch here, hope this one makes it in!

irh commented:
I just tried this out and it's great, nice work! 👏

@drybalka 
To offer my 2¢ on the working vs. buffer directory topic, we already have `file_picker`/`file_picker_in_current_buffer_directory`, so personally I would expect to see `file_browser`/`file_browser_in_current_buffer_directory`, with the `file_browser` picker opening in the current working directory. 

> Not sure when I would even need to open a file browser in the project root. 

This would be useful for me when I'm working with a large project, especially one that I'm not familiar with. e.g. I'm looking at some file deep in the tree and I want to reorient myself and start navigating again from the root.

@peteringram0 
> Im not sure why but there is a flicker when navigating directories

I noticed this too when trying it out with `cargo run`, but the flicker doesn't show for me at all in a release build.

drybalka commented:
@peteringram0 The flicker is an unfortunate limitation of the current picker design. The problem is that I could not find any way to update the currently opened picker, so to switch a folder it actually has to close the current picker with a callback to open a new one in the new folder. Depending on the build profile of your helix (development vs production) and the folder size this may or may not be visible. The solution is not the most elegant, but I was lucky that it exists at all without a major refactoring. I think not so many pickers actually need to update themselves and so this functionality was not yet implemented. Hopefully after the refactoring of the pickers this can be improved.

peteringram0 commented:
> @peteringram0 The flicker is an unfortunate limitation of the current picker design. The problem is that I could not find any way to update the currently opened picker, so to switch a folder it actually has to close the current picker with a callback to open a new one in the new folder. Depending on the build profile of your helix (development vs production) and the folder size this may or may not be visible. The solution is not the most elegant, but I was lucky that it exists at all without a major refactoring. I think not so many pickers actually need to update themselves and so this functionality was not yet implemented. Hopefully after the refactoring of the pickers this can be improved.

Makes sense and I don't think this issue should merge. Thank you for your time on this!

thomasaarholt commented:
@peteringram0 

> I don't think this issue should merge

What did you mean here?

peteringram0 commented:
> @peteringram0
> 
> > I don't think this issue should merge
> 
> What did you mean here?

Sorry i was supposed to write that I don't think the issue with the flicking should prevent the PR getting merged.

irh commented:
@drybalka That's awesome, thanks 🙏

I'm using it like this:

```toml
[keys.normal.space]
"e" = "file_browser"
"E" = "file_browser_in_current_buffer_directory"
```

 (`b` is used for the buffer picker, so I've gone with '`e`' for 'explorer' 😀)
 
 This behaves exactly as I'd like, great work, no notes 👏

Axlefublr commented:
braced myself to try merging this into my fork and... it was seamless? \
clean commit history, no merge conflicts, *or* conflicts with my fork, majestically

thank you very much! <3

although I'd love for helix to have a proper file browser (not powered by a classic picker, I mean), this is still very very useful, and I think should be in core!

m0ar commented:
Hey @drybalka! I wanted to pop in and report this is working really well for me too, and I'm so thankful for your work! :revolving_hearts: 

Hopefully this cupcake will nourish your motivation for the last stretch just in time for the Christmas stocking: :cupcake: 

@the-mikedavis, any blockers left that needs taking care of? :eyes: 
Testers seem happy, last round of review comments seems taken care of by @drybalka very quickly after they were made.

irh commented:
> I think default bindings for the File Explorer makes sense to add:
> 
> * File Explorer in current buffer's directory bounded by default to `Space + e`
> * File Explorer in working directory bounded by default to `Space + E`, because its scope is bigger

@NikitaRevenco I went with `e` for root and `E` for current buffer directory because this follows the pattern of the existing `f`/`F` mappings (`f` for workspace root, `F` for working directory). 

Although to correctly follow the pattern, the default for `E` should really be `file_browser_in_current_directory` (I just prefer using the buffer directory as the smaller scope so I would remap it).

cgahr commented:
How do you feel about listing the content of directories, similar to the tree command? Maybe limiting the depth to a certain level?

Could this be done while keeping the PR as simply as it is? I think this would significantly increase the value of the file browser.

NikitaRevenco commented:
> How do you feel about listing the content of directories, similar to the tree command? Maybe limiting the depth to a certain level?
> 
> Could this be done while keeping the PR as simply as it is? I think this would significantly increase the value of the file browser.

You can already view contents of the directories when you hover over them

![image](https://github.com/user-attachments/assets/e71907c7-792d-417f-b5d3-1dab57b2c7bd)

thomasaarholt commented:
> Could this be done while keeping the PR as simply as it is?

I highly recommend aiming to merge this with the current feature-set, and instead aim for improvements in later PRs.

NikitaRevenco commented:
> > I think default bindings for the File Explorer makes sense to add:
> > 
> > * File Explorer in current buffer's directory bounded by default to `Space + e`
> > * File Explorer in working directory bounded by default to `Space + E`, because its scope is bigger
> 
> @NikitaRevenco I went with `e` for root and `E` for current buffer directory because this follows the pattern of the existing `f`/`F` mappings (`f` for workspace root, `F` for working directory).
> 
> Although to correctly follow the pattern, the default for `E` should really be `file_browser_in_current_directory` (I just prefer using the buffer directory as the smaller scope so I would remap it).

agreed, consistency is important

it would be odd if Helix didn't map file browser to anything since it's such an important feature. i personally use it all the time, i've been using this branch merged into my own fork for weeks now. It's beautiful. The following keymap makes the most sense to match `space.f` and `space.F`:

```toml
space.e = "file_browser_in_current_directory"
space.E = "file_browser_in_current_buffer_directory"
```

irh commented:
> The following keymap makes the most sense to match space.f and space.F:
> space.e = "file_browser_in_current_directory"
> space.E = "file_browser_in_current_buffer_directory"

Maybe I'm missing something here, but going by the docs we currently have:

```
space.f = "file_picker"
space.F = "file_picker_in_current_directory"
```

so I would expect to see:

```
space.e = "file_browser"
space.E = "file_browser_in_current_directory"
```

I don't want to derail this PR though, I'd be happy to see mappings introduced as a followup if needed.

darshanCommits commented:
> so I would expect to see:
> 
> ```
> space.e = "file_browser"
> space.E = "file_browser_in_current_directory"
> ```

it may not be more consistent as compared to
```space.E = "file_browser_in_current_directory"```
but surely is more useful as the afaik only way to change `current_directory` after opening helix is via `:cd` (not sure if this PR acts differently, i dont have it merged right now)

summersz commented:
My preference is for @NikitaRevenco's suggestion.

I have been using this for a while and exclusively with file_browser_in_current_buffer_directory.  This implementation is similar to netrw which opens at the buffer directory, which I think this is most suited for.

This is still inline with f/F, where the capitalization opens a narrower scope.  I think the browser is more useful at working and buffer directories so makes it sense to deviate.

drybalka commented:
@m0ar Thank you for the cupcake! 

I deliberately did not add the default key mappings because they are often a matter of taste. I would prefer to get a green flag from the core maintainers with a wider vision for helix first, or maybe one of them may even add a commit with the mappings themself before the merge, as it may be faster than writing a github comment =) I personally use `space+e/E` for the "explorer" mnemonic, but any other combination will do as well.

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -265,6 +266,76 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
+    let mut dirs = Vec::new();
+    let mut files = Vec::new();
+    for entry in std::fs::read_dir(path)?.flatten() {
+        if entry.path().is_dir() {
```
We should try to avoid `Path::is_dir` (and equivalents like `PathBuf::is_dir`) since they mean a syscall per call. We can avoid that here (most of the time, platform dependent) by using `DirEntry::file_type`:

```suggestion
        if entry.file_type().is_ok_and(|file_type| file_type.is_dir()) {
```

I would also switch the return Vec from this function to be `(PathBuf, bool)`, so we keep that `is_dir` boolean around rather than calling `path.is_dir()` elsewhere (especially while rendering). This will make the file browser type different than FilePicker:

```rust
type FileBrowser = Picker<(PathBuf, bool), PathBuf>;
```

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -265,6 +266,76 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
+        |item: &PathBuf, root: &PathBuf| {
+            let name = item.strip_prefix(root).unwrap_or(item).to_string_lossy();
+            if item.is_dir() {
+                format!("{}/", name).into()
```
Building on the changes in #12205 we can style directories with `"ui.text.directory"`. So `file_browser` will need to be passed `&Editor` to read from the theme and then the Picker's data can be a tuple or a new struct that holds the root and the directory style. So this column definition would instead become something like:

```rust
let columns = [PickerColumn::new(
    "path",
    |(path, is_dir): &(PathBuf, bool), data: &FileBrowserData| {
        let name = path
            .strip_prefix(&data.root)
            .unwrap_or(path)
            .to_string_lossy();
        if *is_dir {
            Span::styled(format!("{}/", name), data.directory_style).into()
        } else {
            name.into()
        }
    },
)];
```

the-mikedavis replied:
And this styling can be applied in the Picker rendering function so that directories are also `"ui.text.directory"` in the preview

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -265,6 +266,76 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
 }
 
+pub fn file_browser(root: PathBuf) -> Result<FilePicker, std::io::Error> {
+    let root = helix_stdx::path::canonicalize(root);
```
I'm fairly certain we don't need to canonicalize this path. Canonicalization is almost never the right thing to do and it usually causes weird effects with symlinks.

drybalka commented:
@the-mikedavis Thank you very much for the review, I really appreciate the insights! I added the styling for dirs, as you suggested, and also removed canonicalization and unnecessary is_dir syscalls. Now the type signatures are a bit polluted with `(PathBuf, bool)` tuples, but I just couldn't come up with a decent name for it (PathBufWithSomeMetadata or FileBrowserEntry are not better, I think). Hopefully this is fine.

Have a Merry Christmas!

Axlefublr commented:
thank you for rebasing @drybalka! for us forkers that is very valuable :3 \
hope this PR will get merged into core anyway, though!

NikitaRevenco commented on helix-term/src/commands.rs:
```diff
@@ -2986,6 +2989,58 @@ fn file_picker_in_current_directory(cx: &mut Context) {
+                );
+                return;
+            }
+            cx.editor.set_error(
```
I don't think this should be a displayed error. For example I'm frequently doing `hx .` and then opening the file browser, since i am in a `[scratch]`, everytime I do that it gives me this error, even though this isn't really an error

drybalka replied:
To be fair, this function is called `file_browser_in_current_buffer_directory`, so this error is quite justified to inform the user that the `[scratch]` buffer does not have a dedicated directory. In my opinion, it would be confusing to remove this message for cases when something actually goes wrong.

What you want to call after `hx .` is probably `file_browser_in_current_directory` or simply `file_browser`. Now I agree, that it is more convenient to have a single key combination that covers all file browsing needs in one go, and arguably `file_browser_in_current_buffer_directory` already does that for the price of showing you a somewhat-irritating-but-easily-ignorable error message. In an ideal world one would probably have to wait for when programmable config will be available in helix and code the desired behavior themself.

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -276,6 +279,74 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
+        (root, directory_style),
+        move |cx, (path, is_dir): &(PathBuf, bool), action| {
+            if *is_dir {
+                let owned_path = path.clone();
```
```suggestion
                let new_root = helix_stdx::path::normalize(path);
```

when switching paths we should normalize the input path because it may be `path.join("..")` and `owned_path.parent()` returns `path` (https://play.rust-lang.org/?version=stable&mode=debug&edition=2021). The effect is that if you use the file browser to try to reach the file system root then you'll still see `../` in the options unless we normalize the new root here

drybalka replied:
Yes, that was the reason why I canonicalized the path before =) I was not aware of the normalize utility function, thanks!

the-mikedavis commented on helix-term/src/commands.rs:
```diff
@@ -393,6 +393,9 @@ impl MappableCommand {
         file_picker, "Open file picker",
         file_picker_in_current_buffer_directory, "Open file picker at current buffer's directory",
         file_picker_in_current_directory, "Open file picker at current working directory",
+        file_browser, "Open file browser in workspace root",
```
I think it would be nice to set a default keybind of `<space>.` for this. Thoughts @archseer? @pascalkuthe?

pascalkuthe replied:
Seems fine to me but I haven't looked at this PR much yet

RoloEdits replied:
I think `<space>e` could also work. Could have a `<space>E`, similar in nature to the file picker.

KevinDanne replied:
I have created a PR that adds mappings and fixes the docs check https://github.com/drybalka/helix/pull/1

civa commented:
> Actually, I can't get this to work at all? Running on MacOS.
> 
>  Screen.Recording.2024-09-10.at.08.47.35.mov 
> 
> [Here is the log](https://gist.github.com/thomasaarholt/fd8de460bb7bf8d028bd7038421ae6d1) from `hx -vvv .`

Same issue

nik-rev commented:
> > Actually, I can't get this to work at all? Running on MacOS.
> > Screen.Recording.2024-09-10.at.08.47.35.mov
> > [Here is the log](https://gist.github.com/thomasaarholt/fd8de460bb7bf8d028bd7038421ae6d1) from `hx -vvv .`
> 
> Same issue

You can't open *any* pickers from the command line, it isn't limited to just this PR. It's a known bug for a while and fixing it would require a large refactor. See https://github.com/helix-editor/helix/issues/4508

nik-rev commented:
It's merged!! Yayyyyyyy :D

(small thing, but I noticed it's mapped to `space` + `e` but the command is called `file_browser`. maybe it's better to rename it to `file_explorer` as a mnemonic?)

Axlefublr commented:
NO FUCKING WAY FINALLY OMG

the-mikedavis commented:
It does look a bit like Vim's `:Explore`, "explore" could be a more fitting name

rotmh commented:
Hi, awesome and exciting! :)

One thing, it seems not to ignore ignored files (from `.gitignore`, `.ignore`, etc.). Is that the desired behavior, a problem in my case, maybe it has been forgotten?

DoctorRyner commented:
The next step would be file operations, like creating and removing files

rotmh commented:
Like `oil.nvim`?

Axlefublr commented:
oil.nvim is way out of scope for this \
or even file operations I'd argue

let's not forget that to open a directory, a picker is *re*created \
I think it's silly to build on a base this flimsy

I'd just see this is an alternative for a file *picker*, but nothing more

DoctorRyner commented:
> Like `oil.nvim`?

More like https://github.com/nvim-telescope/telescope-file-browser.nvim that was mentioned on the PR

rotmh commented:
Thanks :)

That's what I thought, as `oil.nvim` is definitely plugin material (and not core material), but this

> The next step would be file operations, like creating and removing files

made me second that.

Anyway, I'm more of a picker kinda person for moving between files, but having a comfortable solution for batch files/dirs editing in-editor was something I liked in `nvim` + `oil`.

DoctorRyner commented:
@rotmh ye, I use VS Code for working with files, moving scripts that automatically updates imports. But it looks weird. The only 2 things I use VS Code for are file management and merge conflicts. Weeeell, easy conflicts I can resolve with space + g when there are not a lot of them, but it's no good if file is huge and there are lots of conflicts

Misnad commented:
Is it possible to use vim motions in file picker (h,j,k,l for left,down,up,right)?
